File: verification/decision/engine.go
package decision

import (
	"fmt"
	"strings"

	"doctor-service/entities"
	"doctor-service/verification/comparison"
	"doctor-service/verification/council"
	"doctor-service/verification/fraud"
)

type DecisionResult struct {
	FinalStatus entities.DoctorStatus `json:"final_status"`
	Reason      string                `json:"reason"`
}

type DecisionEngine struct{}

func NewDecisionEngine() *DecisionEngine {
	return &DecisionEngine{}
}

func (e *DecisionEngine) Evaluate(
	comp *comparison.ComparisonResult,
	councilRes *council.CouncilVerificationResult,
	fraudRes *fraud.FraudAnalysisResult,
	ocrConfidence float64,
) *DecisionResult {
	// 1. REJECTION Rules
	if councilRes != nil && councilRes.Status == "SUSPENDED" {
		return &DecisionResult{
			FinalStatus: entities.DoctorStatusRejected,
			Reason:      "Automatic rejection: Medical registration is blacklisted or suspended by the state/national council.",
		}
	}

	if fraudRes.FraudScore >= 75 {
		return &DecisionResult{
			FinalStatus: entities.DoctorStatusRejected,
			Reason:      fmt.Sprintf("Automatic rejection: Critical fraud risk score detected (%d/100).", fraudRes.FraudScore),
		}
	}

	// 2. AUTO_VERIFIED Rules
	if councilRes != nil && councilRes.Status == "ACTIVE" &&
		comp.RegNoMatch &&
		comp.NameMatchScore >= 90.0 &&
		ocrConfidence >= 85.0 &&
		fraudRes.FraudScore <= 15 {

		return &DecisionResult{
			FinalStatus: entities.DoctorStatusAutoVerified,
			Reason:      "Automated verification successful: Medical council registration verified active, high document OCR confidence, and low risk score.",
		}
	}

	// 3. MANUAL_REVIEW Rules (Default for discrepancies or moderate confidence)
	reasons := []string{}
	if comp.NameMatchScore < 90.0 {
		reasons = append(reasons, fmt.Sprintf("Name similarity score %.1f%% below auto-verify threshold (90%%)", comp.NameMatchScore))
	}
	if !comp.RegNoMatch {
		reasons = append(reasons, "Registration number on certificate did not match submitted registration number")
	}
	if ocrConfidence < 85.0 {
		reasons = append(reasons, fmt.Sprintf("OCR confidence %.1f%% below auto-verify threshold (85%%)", ocrConfidence))
	}
	if councilRes == nil || councilRes.Status != "ACTIVE" {
		reasons = append(reasons, "Medical council online registry lookup required manual confirmation")
	}
	if fraudRes.FraudScore > 15 {
		reasons = append(reasons, fmt.Sprintf("Elevated risk score (%d/100)", fraudRes.FraudScore))
	}

	reasonStr := "Flagged for manual reviewer approval."
	if len(reasons) > 0 {
		reasonStr = fmt.Sprintf("Flagged for manual review: %s.", strings.Join(reasons, "; "))
	}

	return &DecisionResult{
		FinalStatus: entities.DoctorStatusManualReview,
		Reason:      reasonStr,
	}
}

File: backend_api/verification/decision/engine.py
from typing import Optional
from ..comparison.comparator import ComparisonResult
from ..council.provider import CouncilVerificationResult
from ..fraud.detector import FraudAnalysisResult

class DecisionResult:
    def __init__(self, final_status: str, reason: str):
        self.final_status = final_status
        self.reason = reason

class DecisionEngine:
    def evaluate(
        self,
        comp: ComparisonResult,
        council_res: Optional[CouncilVerificationResult],
        fraud_res: FraudAnalysisResult,
        ocr_confidence: float
    ) -> DecisionResult:
        # 1. REJECTION Rules
        if council_res is not None and council_res.status == "SUSPENDED":
            return DecisionResult(
                final_status="REJECTED",
                reason="Automatic rejection: Medical registration is blacklisted or suspended by the state/national council."
            )

        if fraud_res.fraud_score >= 75:
            return DecisionResult(
                final_status="REJECTED",
                reason=f"Automatic rejection: Critical fraud risk score detected ({fraud_res.fraud_score}/100)."
            )

        # 2. AUTO_VERIFIED Rules
        if (council_res is not None and council_res.status == "ACTIVE" and
            comp.reg_no_match and
            comp.name_match_score >= 90.0 and
            ocr_confidence >= 85.0 and
            fraud_res.fraud_score <= 15):
            
            return DecisionResult(
                final_status="AUTO_VERIFIED",
                reason="Automated verification successful: Medical council registration verified active, high document OCR confidence, and low risk score."
            )

        # 3. MANUAL_REVIEW Rules (Default for discrepancies or moderate confidence)
        reasons = []
        if comp.name_match_score < 90.0:
            reasons.append(f"Name similarity score {comp.name_match_score:.1f}% below auto-verify threshold (90%)")
        if not comp.reg_no_match:
            reasons.append("Registration number on certificate did not match submitted registration number")
        if ocr_confidence < 85.0:
            reasons.append(f"OCR confidence {ocr_confidence:.1f}% below auto-verify threshold (85%)")
        if council_res is None or council_res.status != "ACTIVE":
            reasons.append("Medical council online registry lookup required manual confirmation")
        if fraud_res.fraud_score > 15:
            reasons.append(f"Elevated risk score ({fraud_res.fraud_score}/100)")

        reason_str = "Flagged for manual reviewer approval."
        if reasons:
            reason_str = f"Flagged for manual review: {'; '.join(reasons)}."

        return DecisionResult(
            final_status="MANUAL_REVIEW",
            reason=reason_str
        )
